deepspeed/comm/comm.py

```diff
@@ -336,16 +334,16 @@ def allgather_fn(output_tensor,
     global cdb
     global has_warned_all_gather
     assert cdb is not None and cdb.is_initialized(), 'DeepSpeed backend not set, please initialize it using init_process_group()'
-    if cdb.has_allgather_base:
-        return all_gather_base(output_tensor,
-                               input_tensor,
-                               group=group,
-                               async_op=async_op,
-                               debug=debug)
+    if cdb.has_all_gather_into_tensor():
```

Comment:
This is not backwards-compatible with older versions of torch that should still be using `all_gather_base`. I think a better structure would be:

```
if cdb.has_all_gather_into_tensor():
    return all_gather_into_tensor(...)
elif cdb.has_allgather_base():
    return all_gather_base(...)
else:
    # print warning on all_gather perf
    return all_gather(...)
```

it'd be cleaner to use your `get_all_gather_function()` here and only print the warning if `get_all_gather_function is None`

Reply:
I think it should work with older versions too right?
<img width="663" alt="Screenshot 2023-03-15 at 3 27 15 PM" src="https://user-images.githubusercontent.com/32954280/225274174-e5c2115b-4072-4745-a5ff-0fcad429aa18.png">
this populates the function and for older versions it should work.

Am i missing something?

Reply:
I am populating the all_gather_function here:
https://github.com/mayank31398/DeepSpeed/blob/3133caff25307d63dc248b02416420eb9288189d/deepspeed/comm/torch.py#L28

Reply:
Ah I see. The naming of `has_all_gather_into_tensor` is a bit confusing since this checks for both `all_gather_into_tensor` and `_all_gather_base`. 

I think https://github.com/microsoft/DeepSpeed/pull/3020 introduces a more clear naming convention with `allgather_base_fn`

Reply:
I agree :)

Reply:
@mayank31398, do you plan to adapt the naming convention to #3020? I am fine either way.

Reply:
not really, the new convention in torch is `gather_into_tensor`, so it would be better if we adopt this sooner?